File: src/main/java/xyz/upperlevel/hgame/scenario/Event.java
package xyz.upperlevel.hgame.scenario;

import xyz.upperlevel.hgame.scenario.character.Actor;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map.Entry;
import java.util.function.Supplier;

public class Event {
    private static final List<Entry<Supplier<Boolean>, Runnable>> events = new ArrayList<>();

    public static void watch(Supplier<Boolean> trigger, Runnable action) {
        events.add(new AbstractMap.SimpleEntry<>(trigger, action));
    }

    public static void update() {
        List<Entry<Supplier<Boolean>, Runnable>> eventsCopy = new ArrayList<>(events);
        for (Entry<Supplier<Boolean>, Runnable> entry : eventsCopy) {
            if (entry.getKey().get()) {
                entry.getValue().run();
                events.remove(entry);
            }
        }
    }

    private Event() {
    }

    public static Supplier<Boolean> greaterX(Actor actor, float x) {
        return () -> actor.getX() >= x;
    }

    public static Supplier<Boolean> lowerX(Actor actor, float x) {
        return () -> actor.getX() <= x;
    }
}

File: src/main/java/xyz/upperlevel/hgame/world/Event.kt
package xyz.upperlevel.hgame.world

import xyz.upperlevel.hgame.world.character.Entity
import java.util.*
import java.util.function.Supplier
import kotlin.collections.Map.Entry

object Event {
    private val events = ArrayList<Entry<Supplier<Boolean>, Runnable>>()

    fun watch(trigger: Supplier<Boolean>, action: Runnable) {
        events.add(AbstractMap.SimpleEntry(trigger, action))
    }

    fun update() {
        val eventsCopy = ArrayList(events)
        for (entry in eventsCopy) {
            if (entry.key.get()) {
                entry.value.run()
                events.remove(entry)
            }
        }
    }

    fun greaterX(entity: Entity, x: Float): () ->  Boolean {
        return { entity.x >= x }
    }

    fun lowerX(entity: Entity, x: Float): () -> Boolean {
        return { entity.x <= x }
    }
}
